lms/djangoapps/completion_api/models.py

```diff
@@ -0,0 +1,249 @@
+    """
+    Facade wrapping progress.models.StudentProgress model.
+    """
+
+    def __init__(self, inner):
+        self._blocks = None
+        self._collected = None
+        self._completable_blocks = None
+        self._inner = inner
+
+    @property
+    def collected(self):
+        """
+        Return the collected block structure for this course
+        """
+        if self._collected is None:
+            self._collected = get_course_in_cache(self.course_key)
+        return self._collected
+
+    @property
+    def blocks(self):
+        """
+        Return all blocks in the course.
+        """
+        if self._blocks is None:
+            course_location = CourseOverview.load_from_module_store(self.course_key).location
+            self._blocks = get_course_blocks(
+                self.user,
+                course_location,
+                collected_block_structure=self.collected,
+            )
+        return self._blocks
+
+    @property
+    def completable_blocks(self):
+        """
+        Return the set of blocks that can be completed that are visible to
+        self.user.
+
+        This method encapsulates the facade's access to the modulestore, making
+        it a useful candidate for mocking.
+        """
+        if self._completable_blocks is None:
+            self._completable_blocks = {
+                block for block in self.blocks
+                if self.blocks.get_xblock_field(block, 'category') not in IGNORE_CATEGORIES
+            }
+        return self._completable_blocks
+
+    @property
+    def user(self):
+        """
+        Return the StudentProgress user
+        """
+        return self._inner.user
+
+    @property
+    def course_key(self):
+        """
+        Return the StudentProgress course_key
+        """
+        return self._inner.course_id
+
+    @property
+    def earned(self):
+        """
+        Return the number of completions earned by the user.
+        """
+
+        return self._inner.completions
+
+    @property
+    def possible(self):
+        """
+        Return the maximum number of completions the user could earn in the
+        course.
+        """
+
+        return float(len(self.completable_blocks))
+
+    @property
+    def percent(self):
+        """
+        Return the percent of the course completed by the user.
+
+        Percent is returned as an integer between 0 and 100.
+        """
+        return int(round(100 * self.earned / self.possible))
+
+    def get_blocks_by_category(self, category):
+        """
+        Return all blocks of the specified category
+        """
+        for block in self.blocks:
+            if self.blocks.get_xblock_field(block, 'category') == category:
+                yield block
+
+    @property
+    def subsections(self):
+        """
+        Return the BlockCompletion for all subsections.
+        """
+        subsections = []
+        for block in self.get_blocks_by_category('sequential'):
```

Comment:
Yes.  I got tripped up on this a few times when working with it.  The function claims to return blocks, so I was looking for a key field, found an id field, tried to use it, and it turned out to be a string.  So block_key will be a better term (and less misleading for other programmers who find the code and try to learn from it).